src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+
+            IDisposable projectLink = ProjectDataSources.SyncLinkTo(
+                   _inputsDataSource.SourceBlock.SyncLinkOptions(
+                       linkOptions: new StandardRuleDataflowLinkOptions
+                       {
+                           PropagateCompletion = true,
+                       }),
+                   _projectSubscriptionService.ProjectRuleSource.SourceBlock.SyncLinkOptions(
+                      linkOptions: new StandardRuleDataflowLinkOptions
+                      {
+                          RuleNames = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName),
+                          PropagateCompletion = true,
+                      }),
+                   _inputsActionBlock,
+                   DataflowOption.PropagateCompletion,
+                   cancellationToken: _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+
+            // Create an action block to process file change notifications
+            _fileWatcherActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<string[]>>(ProcessFileChangeNotification);
+            IDisposable watcherLink = _fileWatcher.SourceBlock.LinkTo(_fileWatcherActionBlock, DataflowOption.PropagateCompletion);
+
+            _disposables.AddDisposable(projectLink);
+            _disposables.AddDisposable(watcherLink);
+
+            return Task.CompletedTask;
+        }
+
+        internal void ProcessFileChangeNotification(IProjectVersionedValue<string[]> arg)
+        {
+            // Ignore any updates until we've received the first set of design time inputs (which shouldn't happen anyway)
+            // That first update will queue all of the files so we're not losing anything
+            if (_designTimeInputs == null)
+            {
+                return;
+            }
+
+            foreach (string changedFile in arg.Value)
+            {
+                // if a shared input changes, we recompile everything
+                if (_sharedDesignTimeInputs.Contains(changedFile))
+                {
+                    foreach (string file in _designTimeInputs)
+                    {
+                        ImmutableInterlocked.TryAdd(ref _filesToCompile, file, /* ignoreWriteTime */ false);
+                    }
+                    // Since we've just queued every file, we don't care about any other changes
+                    break;
+                }
+                else
+                {
+                    // A normal design time input, so just add it to the queue
+                    ImmutableInterlocked.TryAdd(ref _filesToCompile, changedFile, /* ignoreWriteTime */ false);
+                }
+            }
+
+            QueueCompilation();
+        }
+
+        internal void ProcessDataflowChanges(IProjectVersionedValue<Tuple<DesignTimeInputs, IProjectSubscriptionUpdate>> input)
```

Comment:
I would consider having DesignTimeInputs push "full paths" from metadata rather than having this class "MakeRooted" on the result. On top of that, this code is de-duping the paths, but doing on a non-canonical path, which means that the following won't be de-dup'd:

```
<Foo Include="...\Foo.cs"/>
<Foo Include="C:\Project\Foo.cs"/>
```

The other added benefit is that the string will be shared for other consumers that use FullPath.

Reply:
We should do the same everywhere - just have everyone work off FullPath so that we start sharing the strings.

Reply:
I'll have to think about this. At the moment compilation is the only thing in the chain that cares about full paths, so I'm not sure whether using full paths will just mean the other blocks needing to call `MakeRelative` all the time, eg to return the list of inputs, fire change events, subscribe to file watching etc. I think there would be more sharing happening now.

Your point about de-duping the wrong things is valid, though its worth pointing out that anything with a path like in your example will be considered as a linked file, so isn't a valid DesignTimeInput anyway. I need to think about whether there is a way to express the same file with different monikers and have them both be valid inputs. I also think the risk is pretty low, as duplicate files in the compilation queue mean after the first entry gets compiled, the second entry wouldn't pass the file write check (though granted, sometimes that is skipped).

Reply:
>  though its worth pointing out that anything with a path like in your example will be considered as a linked file, 

That's only for files outside of the project cone. Regardless, you can write relative paths in many ways.